### ISSUE TITLE
Bug: Data base not found for database name containing plus sign

### ISSUE THREAD

**Issue by mehrdadep** [gesellix/couchdb-prometheus-exporter#84]
The bug is possibly occurred in `getDatabasesStatsByDbName` function. 

- `numbers/+4485` becomes `numbers%2F+4485` instead of `numbers%2F%2B9809` 

This will cause this error on pulling metrics:

```
E1101 09:41:47.140749   18781 collector.go:422] Error collecting stats: error collecting couchdb stats: error reading database 'numbers/+4485' stats: status 404 Object Not Found (404): {"error":"not_found","reason":"Database does not exist."}
```

**Comment by gesellix** [comment]
Thanks for your bug report! Before reviewing your pull request, I'd like to collect some details about the underlying issue:

According to [RFC 3986, section 2.2](https://tools.ietf.org/html/rfc3986#section-2.2), we have to consider `+` as reserved character as element of the "sub-delims" set. Depending on the context, such characters might have to be encoded. Golang's implementation of [url.PathEscape](https://golang.org/pkg/net/url/#PathEscape) refers to [RFC 3986, section 3.3](https://tools.ietf.org/html/rfc3986#section-3.3), where a path segment is allowed to contain "sub-delims" in _non-encoded_ form.

So, from perspective of Golang following the RFC, we shouldn't have to change anything, because `+` doesn't have to be encoded in path segments.

Looking at CouchDB's implementation things become nasty. Quoting https://github.com/apache/couchdb/issues/825#issuecomment-330311639:

> Yap, but mochiweb decodes + in path as a space, probably because it reuses the code used for application/x-www-form-urlencoded. So if we switched to a different server the code the converting + to space would never be called.

The related change in https://github.com/apache/couchdb/pull/828 shows that `+`, `%20`, and `%2B` are treated same.

I tried to create a database with `+` in its name, with a strange result (same behaviour on CouchDB 1.7.2 and 2.3.1):

```
$ curl -v -X PUT 'http://localhost:5984/foo+234'
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 5984 (#0)
> PUT /foo+234 HTTP/1.1
> Host: localhost:5984
> User-Agent: curl/7.64.1
> Accept: */*
> 
< HTTP/1.1 400 Bad Request
< Cache-Control: must-revalidate
< Content-Length: 198
< Content-Type: application/json
< Date: Sun, 01 Nov 2020 10:30:00 GMT
< Server: CouchDB/2.3.1 (Erlang OTP/19)
< X-Couch-Request-ID: e785047d32
< X-CouchDB-Body-Time: 0
< 
{"error":"illegal_database_name","reason":"Name: 'foo 234'. Only lowercase characters (a-z), digits (0-9), and any of the characters _, $, (, ), +, -, and / are allowed. Must begin with a letter."}
* Connection #0 to host localhost left intact
* Closing connection 0
```

Creating a database `foo+234` via the Fauxton web ui worked, though.

To summarize:
- Golang's implementation (encoding path segments) keeps `+`, which follows common standards (RFC 3986).
- The CouchDB Prometheus exporter currently uses Golang's default implementation.
- CouchDB's implementation allows `+` in database names, but has special handling and actually requires encoding, mostly for, let's say "historical reasons".

I'm hesitating about applying _query_ encoding for path segments, because I'm not sure if other use cases break, but it seems like any CouchDB client has to follow CouchDB's special requirements.

**Comment by gesellix** [comment]
I filed https://github.com/apache/couchdb/issues/3236 for discussion on the CouchDB side. Let's give them a few days to give an answer.

@mehrdadep Which CouchDB version do you currently use?

**Comment by mehrdadep** [comment]
Thank you for this complete explanation about the issue. Our CouchDB version is 2.0.0 and I think since `couchdb-prometheus-exporter` is a client for CouchDB it should be compatible with any version of that. As a side note we're not making these databases from the web ui.

**Comment by gesellix** [comment]
I'm going to merge your pull request this week - you're not forgotten :)

**Comment by gesellix** [comment]
I've just released your fix as v29.3.1, packages and the Docker image are building right now.

In other news: the releated CouchDB issue https://github.com/apache/couchdb/issues/3236 got some feedback. Maybe you're interested in following the discussion over there.

Thanks a lot for your contribution and patience!

**Comment by mehrdadep** [comment]
Yeah sure, thank you
